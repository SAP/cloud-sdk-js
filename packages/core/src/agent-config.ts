import * as http from 'http';

/**
 * Interface for the http-agent within the Axios request config.
 */
export interface HttpAgentConfig {
  httpAgent: http.Agent;
}

/**
 * Interface for the https-agent within the Axios request config.
 */
export interface HttpsAgentConfig {
  httpsAgent: http.Agent;
}
