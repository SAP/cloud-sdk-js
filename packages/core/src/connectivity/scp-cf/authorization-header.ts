import {
  unixEOL,
  isNullish,
  createLogger,
  pickIgnoreCase,
  toSanitizedObject,
  encodeBase64
} from '@sap-cloud-sdk/util';
import type { ODataRequest, ODataRequestConfig } from '../../odata-common';
import {
  AuthenticationType,
  Destination,
  DestinationAuthToken,
  sanitizeDestination
} from './destination';

const logger = createLogger({
  package: 'core',
  messageContext: 'authorization-header'
});

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given ODataRequest to existing headers.
 *
 * @param request - an ODataRequest.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export async function addAuthorizationHeader<
  RequestT extends ODataRequestConfig
>(
  request: ODataRequest<RequestT>,
  headers: Record<string, string>
): Promise<Record<string, string>> {
  const destination = request.destination;
  if (!destination) {
    return headers;
  }
  const authHeaders = await getAuthHeaders(
    destination,
    request.config.customHeaders
  );
  return {
    ...headers,
    ...authHeaders
  };
}

function needsAuthHeaders(destination: Destination): boolean {
  if (!destination.authentication) {
    return false;
  }
  const authTypesWithAuthorizationHeader: AuthenticationType[] = [
    'BasicAuthentication',
    'OAuth2ClientCredentials',
    'OAuth2SAMLBearerAssertion',
    'OAuth2UserTokenExchange',
    'OAuth2JWTBearer',
    'PrincipalPropagation'
  ];
  return authTypesWithAuthorizationHeader.includes(destination.authentication);
}

function getCustomAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
) {
  if (destination.authentication === 'PrincipalPropagation') {
    return pickIgnoreCase(customHeaders, 'SAP-Connectivity-Authentication');
  }
  return pickIgnoreCase(customHeaders, 'authorization');
}

export async function getAuthHeaders(
  destination: Destination,
  customHeaders?: Record<string, any>
): Promise<Record<string, string>> {
  const customAuthHeaders = getCustomAuthHeaders(destination, customHeaders);

  if (Object.keys(customAuthHeaders).length && needsAuthHeaders(destination)) {
    logger.warn(
      'Found custom authorization headers. The given destination also provides authorization headers. This might be unintended. The custom headers from the request config will be used.'
    );
  }

  const additionalDestinationAuthHeaders = getCustomAuthHeaders(
    destination,
    destination.headers
  );

  if (Object.keys(customAuthHeaders).length) {
    return customAuthHeaders;
  }
  if (Object.keys(additionalDestinationAuthHeaders).length) {
    return additionalDestinationAuthHeaders;
  }
  return buildAuthorizationHeaders(destination);
}

/**
 * @deprecated Since v1.20.0. Use [[buildAuthorizationHeaders]] instead.
 * Adds authorization headers for a given destination to existing headers.
 *
 * @param destination - A destination.
 * @param headers - The headers that should be added to.
 * @returns The provided headers with the new authorization headers.
 */
export function buildAndAddAuthorizationHeader(destination: Destination) {
  return async function (
    headers: Record<string, any>
  ): Promise<Record<string, string>> {
    return {
      ...headers,
      ...(await buildAuthorizationHeaders(destination))
    };
  };
}

function toAuthorizationHeader(authorization: string): Record<string, string> {
  return toSanitizedObject('authorization', authorization);
}

function headerFromTokens(
  authenticationType: AuthenticationType,
  authTokens?: DestinationAuthToken[] | null
): Record<string, string> {
  if (!authTokens || !authTokens.length) {
    throw Error(
      `\`AuthenticationType\` is "${authenticationType}", but no auth tokens could be fetched from the destination service.`
    );
  }
  const usableTokens = authTokens.filter(
    (token: DestinationAuthToken) => !token.error
  );

  if (!usableTokens.length) {
    throw Error(
      [
        'The destination tried to provide authorization tokens but failed in all cases. This is most likely due to misconfiguration.',
        'Original error messages:',
        ...authTokens.map(token => token.error)
      ].join(unixEOL)
    );
  }
  const authToken = usableTokens[0];
  return toAuthorizationHeader(authToken.http_header.value);
}

function headerFromBasicAuthDestination(
  destination: Destination
): Record<string, string> {
  if (isNullish(destination.username) || isNullish(destination.password)) {
    throw Error(
      'AuthenticationType is "BasicAuthentication", but "username" and / or "password" are missing!'
    );
  }

  return toAuthorizationHeader(
    basicHeader(destination.username, destination.password)
  );
}

export function basicHeader(username: string, password: string): string {
  return 'Basic ' + encodeBase64(`${username}:${password}`);
}

function headerForPrincipalPropagation(
  destination: Destination
): Record<string, any> {
  const principalPropagationHeader =
    destination?.proxyConfiguration?.headers?.[
      'SAP-Connectivity-Authentication'
    ];
  if (!principalPropagationHeader) {
    throw Error(
      'Principal propagation was selected in destination, but no SAP-Connectivity-Authentication bearer header was added by connectivity service.'
    );
  }
  return {
    'SAP-Connectivity-Authentication': principalPropagationHeader
  };
}

function headerForProxy(destination: Destination): Record<string, any> {
  const proxyAuthHeader =
    destination?.proxyConfiguration?.headers?.['Proxy-Authorization'];

  return proxyAuthHeader ? { 'Proxy-Authorization': proxyAuthHeader } : {};
}

// TODO the proxy header are for OnPrem auth and are now handled correctly and should be removed here
// However this would be a breaking change, since we recommended to use 'NoAuthentication' to achieve principal propagation as a workaround.
// Remove this in v2
function legacyNoAuthOnPremiseProxy(
  destination: Destination
): Record<string, any> {
  logger.warn(
    `You are using 'NoAuthentication' in destination: ${destination.name} which is an OnPremise destination. This is a deprecated configuration, most likely you wanted to set-up 'PrincipalPropagation' so please change the destination property to the desired authentication scheme.`
  );

  let principalPropagationHeader;
  try {
    principalPropagationHeader = headerForPrincipalPropagation(destination);
  } catch (e) {
    logger.warn('No principal propagation header found.');
  }

  return {
    ...headerForProxy(destination),
    ...principalPropagationHeader
  };
}

function getProxyRelatedAuthHeaders(
  destination: Destination
): Record<string, any> {
  if (
    destination.proxyType === 'OnPremise' &&
    destination.authentication === 'NoAuthentication'
  ) {
    return legacyNoAuthOnPremiseProxy(destination);
  }

  // The connectivity service will raise an exception if it can not obtain the 'Proxy-Authorization' and the destination lookup will fail early
  return headerForProxy(destination);
}

async function getAuthenticationRelatedHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  logger.debug(
    `Getting authentication related headers for authentication type: ${destination.authentication}`
  );

  switch (destination.authentication) {
    case null:
    case undefined:
      logger.warn(
        'No authentication type is specified on the destination! Assuming "NoAuthentication".'
      );
      return {};
    case 'NoAuthentication':
    case 'ClientCertificateAuthentication':
      return {};
    case 'OAuth2SAMLBearerAssertion':
    case 'OAuth2UserTokenExchange':
    case 'OAuth2JWTBearer':
    case 'OAuth2ClientCredentials':
      return headerFromTokens(
        destination.authentication,
        destination.authTokens
      );
    case 'BasicAuthentication':
      return headerFromBasicAuthDestination(destination);
    case 'PrincipalPropagation':
      return headerForPrincipalPropagation(destination);
    default:
      throw Error(
        `The destination used "${destination.authentication}" as authentication type which is not supported by the SAP Cloud SDK.`
      );
  }
}

export async function buildAuthorizationHeaders(
  destination: Destination
): Promise<Record<string, string>> {
  const sanitizedDestination = sanitizeDestination(destination);
  return {
    ...(await getAuthenticationRelatedHeaders(sanitizedDestination)),
    ...getProxyRelatedAuthHeaders(sanitizedDestination)
  };
}
