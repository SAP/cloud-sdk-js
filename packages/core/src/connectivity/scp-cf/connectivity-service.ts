import { createLogger, ErrorWithCause } from '@sap-cloud-sdk/util';
import { Protocol } from './protocol';
import { ProxyConfiguration } from './connectivity-service-types';
import {
  AuthenticationType,
  Destination
} from './destination/destination-service-types';
import { EnvironmentAccessor } from './environment-accessor';
import { Service } from './environment-accessor-types';
import { serviceToken } from './token-accessor';
import { decodeJwt, isUserToken, JwtPair } from './jwt';

const logger = createLogger({
  package: 'core',
  messageContext: 'connectivity-service'
});

/**
 * Given a destination and a JWT (required for subscriber destinations), this function will add a proxy configuration to a destination.
 * See also [[ProxyConfiguration]].
 *
 * This function will reject if no connectivity service is bound, no XSUAA service with plan application is bound or the client credentials grant with the XSUAA service fails.
 *
 * @Deprecated Since v1.16.0. Use [[addProxyConfigurationOnPrem]] instead.
 * @param destination - The destination to which the proxy configuration should be added.
 * @param jwt - The JWT of the current user.
 * @returns A promise resolving to the destination with the added proxy configuration.
 */
export function addProxyConfiguration(
  destination: Destination,
  jwt?: string
): Promise<Destination> {
  return Promise.resolve()
    .then(() => proxyHostAndPort())
    .then(hostAndPort =>
      addHeaders(hostAndPort, destination.authentication, jwt)
    )
    .then(proxyConfiguration => ({
      ...destination,
      proxyConfiguration
    }));
}

// TODO: remove string argument in v2.0
export function addProxyConfigurationOnPrem(
  destination: Destination,
  jwt: string | JwtPair | undefined
): Promise<Destination> {
  const jwtPair =
    typeof jwt === 'string' ? { encoded: jwt, decoded: decodeJwt(jwt) } : jwt;
  if (
    destination.authentication === 'PrincipalPropagation' &&
    !isUserToken(jwtPair)
  ) {
    throw new Error('For principal propagation a user JWT is needed.');
  }

  return addProxyConfiguration(destination, jwtPair?.encoded);
}

interface HostAndPort {
  host: string;
  port: number;
  protocol: Protocol.HTTP;
}

export function proxyHostAndPort(): HostAndPort {
  const service = readConnectivityServiceBinding();
  return {
    host: service.credentials.onpremise_proxy_host,
    port:
      service.credentials.onpremise_proxy_http_port ||
      service.credentials.onpremise_proxy_port,
    protocol: Protocol.HTTP
  };
}

function readConnectivityServiceBinding(): Service {
  const serviceBindings = EnvironmentAccessor.getServiceList('connectivity');

  if (!serviceBindings.length) {
    throw new Error(
      'No binding to a connectivity service found! Please make sure to bind an instance of the connectivity service to your app if you want to connect to on-premise destinations.'
    );
  }

  return serviceBindings[0];
}

function addHeaders(
  hostAndPort: HostAndPort,
  authenticationType: AuthenticationType | undefined,
  jwt?: string
): Promise<ProxyConfiguration> {
  const connServiceBinding = readConnectivityServiceBinding();

  return Promise.resolve()
    .then(() => proxyAuthorizationHeader(connServiceBinding, jwt))
    .then(proxyAuthHeader => ({
      ...proxyAuthHeader,
      ...sapConnectivityAuthenticationHeader(authenticationType, jwt)
    }))
    .then(
      headers =>
        ({
          ...hostAndPort,
          headers
        } as ProxyConfiguration)
    );
}

function proxyAuthorizationHeader(
  connectivityServiceBinding: Service,
  userJwt?
): Promise<Record<string, string>> {
  return serviceToken(connectivityServiceBinding, { userJwt })
    .then(token => ({
      'Proxy-Authorization': `Bearer ${token}`
    }))
    .catch(error => {
      throw new ErrorWithCause(
        'Failed to add proxy authorization header - client credentials grant failed!',
        error
      );
    });
}

function sapConnectivityAuthenticationHeader(
  authenticationType: AuthenticationType | undefined,
  jwt?: string
): Record<string, string> {
  if (authenticationType === 'PrincipalPropagation') {
    if (jwt) {
      return {
        'SAP-Connectivity-Authentication': `Bearer ${jwt}`
      };
    }
    throw new Error(
      `Unable to create "SAP-Connectivity-Authentication" header: no JWT found on the current request.
     Connecting to on-premise systems via principle propagation is not possible.`
    );
  }
  if (authenticationType === 'BasicAuthentication') {
    logger.warn(
      'You are connecting to an On-Premise system using basic authentication. For productive usage Principal propagation is recommended.'
    );
  }

  return {};
}
