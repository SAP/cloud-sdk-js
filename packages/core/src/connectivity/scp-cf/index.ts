export * from './jwt';
export * from './cache';
export * from './client-credentials-token';
export * from './client-credentials-token-cache';
export * from './connectivity-service';
export * from './connectivity-service-types';
export * from './destination';
export * from './environment-accessor';
export * from './environment-accessor-types';
export * from './resilience-options';
export * from './token-accessor';
export * from './xsuaa-service';
export * from './xsuaa-service-types';
export * from './protocol';
export * from './authorization-header';
export * from './get-protocol';
export * from './header-builder-for-destination';
export {
  mappingTenantFields,
  Tenant,
  tenantFromJwt,
  RegisteredJWTClaimsTenant,
  isIdenticalTenant
} from './tenant';
export {
  RegisteredJWTClaimsUser,
  Scope,
  mappingUserFields,
  userFromJwt,
  customAttributes
} from './user';
