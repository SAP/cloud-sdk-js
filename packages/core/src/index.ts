export * from './scp-cf';
export * from './util';
export * from './odata/common';
export * from './odata/v2';
export * from './odata/v4';
export * from './header-builder';
export * from './http-agent';
export * from './http-client';
export * from './protocol';
export * from './get-protocol';
