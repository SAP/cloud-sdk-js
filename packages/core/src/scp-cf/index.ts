export * from './cache';
export * from './client-credentials-token';
export * from './client-credentials-token-cache';
export * from './connectivity-service';
export * from './connectivity-service-types';
export * from './destination';
export * from './destination-accessor';
export * from './destination-accessor-types';
export * from './destination-cache';
export * from './destination-selection-strategies';
export * from './destination-service';
export * from './destination-service-types';
export * from './env-destination-accessor';
export * from './environment-accessor';
export * from './environment-accessor-types';
export * from './resilience-options';
export * from './token-accessor';
export * from './vcap-service-destination';
export * from './xsuaa-service';
export * from './xsuaa-service-types';
