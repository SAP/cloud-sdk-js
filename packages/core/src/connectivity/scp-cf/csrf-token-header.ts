import {
  createLogger,
  ErrorWithCause,
  pickIgnoreCase,
  pickNonNullish,
  pickValueIgnoreCase
} from '@sap-cloud-sdk/util';
import { AxiosError } from 'axios';
import { HttpRequestConfig, executeHttpRequest } from '../../http-client';
import { Destination, DestinationNameAndJwt } from '../scp-cf';
import { removeTrailingSlashes } from '../../odata-common/remove-slashes';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return pickNonNullish({
    ...pickIgnoreCase(csrfHeaders, 'x-csrf-token'),
    cookie: buildCookieHeaderValue(
      pickValueIgnoreCase(csrfHeaders, 'set-cookie')
    )
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const fetchHeader = !pickValueIgnoreCase(
    requestConfig.headers,
    'x-csrf-token'
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  // The S/4 does a redirect if the CSRF token is fetched in case the '/' is not in the URL.
  // TODO: remove once https://github.com/axios/axios/issues/3369 is really fixed. Issue is closed but problem stays.
  return executeHttpRequest(destination, appendSlash(axiosConfig))
    .then(response => response.headers)
    .catch(e1 => {
      logger.error(
        new ErrorWithCause(
          'Initial try to fetch CSRF token failed - retry without slash at ',
          e1
        )
      );
      return executeHttpRequest(destination, removeSlash(axiosConfig))
        .then(response => response.headers)
        .catch(e2 => {
          throw new ErrorWithCause(
            'Also second try to fetch SRF token failed - No CSRF token fetched.',
            e2
          );
        });
    });
}

function appendSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (!requestConfig.url!.endsWith('/')) {
    requestConfig.url = `${requestConfig.url}/`;
  }
  return requestConfig;
}

function removeSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (requestConfig.url!.endsWith('/')) {
    requestConfig.url = removeTrailingSlashes(requestConfig.url!);
  }
  return requestConfig;
}

function axiosWorkaround<T extends HttpRequestConfig>(
  error: AxiosError,
  axiosConfig: Partial<T>,
  destination: Destination | DestinationNameAndJwt
) {
  if (error.request?._isRedirect && error.request?._options?.path) {
    logger.warn(
      'Csrf fetch was redirected and failed. This might be a bug in the underlying request library (https://github.com/axios/axios/issues/3369).\nRetrying with full configuration.'
    );
    return makeCsrfRequest(destination, {
      ...axiosConfig,
      url: error.request._options.path
    });
  }
}

function validateCsrfTokenResponse(responseHeaders: Record<string, any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}
