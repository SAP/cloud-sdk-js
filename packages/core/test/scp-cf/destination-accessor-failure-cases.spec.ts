/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import nock from 'nock';
import { clientCredentialsTokenCache, getDestination } from '../../src/scp-cf';
import {
  mockServiceBindings,
  mockXsuaaBinding
} from '../test-util/environment-mocks';
import {
  providerServiceToken,
  subscriberServiceToken,
  subscriberServiceTokenWithVerificationURL,
  subscriberUserJwt,
  userApprovedSubscriberServiceToken
} from '../test-util/mocked-access-tokens';
import {
  mockServiceToken,
  mockUserApprovedServiceToken
} from '../test-util/token-accessor-mocks';
import {
  mockInstanceDestinationsCall,
  mockSingleDestinationCall,
  mockSubaccountDestinationsCall,
  mockVerifyJwt
} from '../test-util/destination-service-mocks';
import {
  basicMultipleResponse,
  destinationName,
  oauthMultipleResponse
} from '../test-util/example-destination-service-responses';

describe('Failure cases', () => {
  beforeEach(() => {
    clientCredentialsTokenCache.clear();
  });

  it('fails if no destination service is bound', async () => {
    process.env['VCAP_SERVICES'] = JSON.stringify({
      xsuaa: [mockXsuaaBinding]
    });

    await expect(
      getDestination(destinationName, {
        userJwt: subscriberServiceTokenWithVerificationURL,
        cacheVerificationKeys: false
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  }, 50000);

  it('throws an error when the provide userJwt is invalid', async () => {
    mockServiceBindings();
    mockVerifyJwt();

    await expect(
      getDestination(destinationName, {
        userJwt: 'fails',
        cacheVerificationKeys: false
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws an error if the subaccount/instance destinations call fails', async () => {
    mockServiceBindings();
    mockVerifyJwt();
    mockServiceToken();

    const httpMocks = [
      mockInstanceDestinationsCall(
        nock,
        {
          ErrorMessage: 'Unable to parse the JWT in Authorization Header.'
        },
        400,
        subscriberServiceToken
      ),
      mockSubaccountDestinationsCall(
        nock,
        basicMultipleResponse,
        200,
        subscriberServiceToken
      )
    ];

    try {
      await getDestination(destinationName, {
        userJwt: subscriberServiceToken,
        enableCircuitBreaker: false,
        cacheVerificationKeys: false
      });
      fail();
    } catch (error) {
      expect(error.message).toContain('Failed to fetch instance destinations');
      expect(error.stack).toContain('status code 400');
      httpMocks.forEach(mock => expect(mock.isDone()).toBe(true));
    }
  });

  it('returns an error if the single destination call fails for OAuth2SAMLBearerAssertion destinations', async () => {
    mockServiceBindings();
    mockVerifyJwt();
    mockServiceToken();
    mockUserApprovedServiceToken();

    const httpMocks = [
      mockInstanceDestinationsCall(nock, [], 200, subscriberServiceToken),
      mockSubaccountDestinationsCall(
        nock,
        oauthMultipleResponse,
        200,
        subscriberServiceToken
      ),
      mockInstanceDestinationsCall(nock, [], 200, providerServiceToken),
      mockSubaccountDestinationsCall(nock, [], 200, providerServiceToken),
      mockSingleDestinationCall(
        nock,
        {
          ErrorMessage: 'Unable to parse the JWT in Authorization Header.'
        },
        401,
        destinationName,
        userApprovedSubscriberServiceToken
      )
    ];

    try {
      await getDestination(destinationName, {
        userJwt: subscriberUserJwt,
        enableCircuitBreaker: false,
        cacheVerificationKeys: false
      });
      fail();
    } catch (error) {
      expect(error instanceof Error).toBeTruthy();
      expect(error.message).toContain(
        'Failed to fetch destination FINAL-DESTINATION'
      );
      expect(error.stack).toContain('status code 401');
      httpMocks.forEach(mock => expect(mock.isDone()).toBe(true));
    }
  });

  it('returns null if no destinations are found', async () => {
    mockServiceBindings();
    mockVerifyJwt();
    mockServiceToken();

    const httpMocks = [
      mockInstanceDestinationsCall(nock, [], 200, subscriberServiceToken),
      mockSubaccountDestinationsCall(nock, [], 200, subscriberServiceToken),
      mockInstanceDestinationsCall(nock, [], 200, providerServiceToken),
      mockSubaccountDestinationsCall(nock, [], 200, providerServiceToken)
    ];

    const expected = null;
    const actual = await getDestination(destinationName, {
      userJwt: subscriberUserJwt,
      cacheVerificationKeys: false
    });
    expect(actual).toEqual(expected);
    httpMocks.forEach(mock => expect(mock.isDone()).toBe(true));
  });

  it('should throw an error when neither userJwt nor SystemUser are defined', async () => {
    mockServiceBindings();
    mockVerifyJwt();
    mockServiceToken();

    const instanceDestinationCallMock = mockInstanceDestinationsCall(
      nock,
      oauthMultipleResponse,
      200,
      providerServiceToken
    );
    const subaccountDestinationCallMock = mockSubaccountDestinationsCall(
      nock,
      [],
      200,
      providerServiceToken
    );

    await expect(
      getDestination(destinationName, { cacheVerificationKeys: false })
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(instanceDestinationCallMock.isDone()).toBe(true);
    expect(subaccountDestinationCallMock.isDone()).toBe(true);
  });
});
