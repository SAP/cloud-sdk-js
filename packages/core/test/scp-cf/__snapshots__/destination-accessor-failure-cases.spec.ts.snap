// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Failure cases fails if no destination service is bound 1`] = `"Failed to verify JWT - unable to get verification key!"`;

exports[`Failure cases should throw an error when neither userJwt nor SystemUser are defined 1`] = `"No user token (JWT) has been provided! This is strictly necessary for principal propagation."`;

exports[`Failure cases throws an error when the provide userJwt is invalid 1`] = `"JwtError: The given jwt payload does not encode valid JSON."`;
